packages/vscode-extension/src/builders/PlatformBuildCache.ts

```diff
@@ -0,0 +1,152 @@
+import { AndroidBuildResult } from "./buildAndroid";
+import { getLaunchConfiguration } from "../utilities/launchConfiguration";
+import { runFingerprintScript } from "./customBuild";
+import { CancelToken } from "./cancelToken";
+import { calculateMD5 } from "../utilities/common";
+import { BuildResult } from "./BuildManager";
+
+const ANDROID_BUILD_CACHE_KEY = "android_build_cache";
+const IOS_BUILD_CACHE_KEY = "ios_build_cache";
+
+const IGNORE_PATHS = [
+  path.join("android", ".gradle/**/*"),
+  path.join("android", "build/**/*"),
+  path.join("android", "app", "build/**/*"),
+  path.join("ios", "build/**/*"),
+  "**/node_modules/**/android/.cxx/**/*",
+  "**/node_modules/**/.gradle/**/*",
+  "**/node_modules/**/android/build/intermediates/cxx/**/*",
+];
+
+export type BuildCacheInfo = {
+  fingerprint: string;
+  buildHash: string;
+  buildResult: AndroidBuildResult | IOSBuildResult;
+};
+
+export class PlatformBuildCache {
+  private cacheKey: string;
+
+  constructor(
+    private readonly platform: DevicePlatform,
+    private readonly cancelToken: CancelToken
+  ) {
+    this.cacheKey =
+      platform === DevicePlatform.Android ? ANDROID_BUILD_CACHE_KEY : IOS_BUILD_CACHE_KEY;
+  }
+
+  public async storeBuild(build: BuildResult) {
+    const fingerprint = await this.calculateFingerprint();
+    const appPath = await getAppHash(getAppPath(build));
+    await extensionContext.workspaceState.update(this.cacheKey, {
+      fingerprint,
+      buildHash: appPath,
+      buildResult: build,
+    });
+  }
+
+  public async clearCache() {
+    await extensionContext.workspaceState.update(this.cacheKey, undefined);
+  }
+
+  public async getBuild() {
+    const cache = extensionContext.workspaceState.get<BuildCacheInfo>(this.cacheKey);
+    if (!cache) {
+      Logger.debug("No cached build found.");
+      return undefined;
+    }
+
+    const currentFingerprint = await this.calculateFingerprint();
+    const fingerprintsMatch = cache.fingerprint === currentFingerprint;
+    if (!fingerprintsMatch) {
+      Logger.info(
+        `Fingerprint mismatch, cannot use cached build. Old: '${cache.fingerprint}', new: '${currentFingerprint}'.`
+      );
+      return undefined;
+    }
+
+    const build = cache.buildResult;
+    const appPath = getAppPath(build);
+    try {
+      const builtAppExists = fs.existsSync(appPath);
+      if (!builtAppExists) {
+        Logger.info("Couldn't use cached build. App artifact not found.");
+        return undefined;
+      }
+
+      const appHash = await getAppHash(appPath);
+      const hashesMatch = appHash === cache.buildHash;
+      if (hashesMatch) {
+        Logger.info("Using cached build.");
+        return build;
+      }
+    } catch (e) {
+      // we only log the error and ignore it to allow new build to start
+      Logger.error("Error while attempting to load cached build: ", e);
+      return undefined;
+    }
+  }
+
+  public async isCacheInvalidated() {
```

Comment:
"invalidating" here is not the right term. It refers to the fact of cleaning the cached data. In this case, it'd be checking if the stored cache key still exists. This method however, checks whether the stored cache is up to date. I'd rename to `hasUpToDateBuild` or something along these lines